apis/secretsmanager/v1alpha1/secret_types.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+// Keys used in connection secret.
+const (
+	ConnectionSecretKeyName        = "secret"
+	ConnectionSecretKeyProjectName = "projectName"
+)
+
+// SecretParameters defines parameters for a desired Secret Manager's secret.
+type SecretParameters struct {
+
+	// Required. The resource name of the project to associate with the
+	// [Secret][google.cloud.secretmanager.v1.Secret], in the format `projects/*`.
+	Parent string `json:"parent,omitempty"`
+
+	// Labels are used as additional metadata on Topic.
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// The replication policy of the secret data attached to the Secret. Can be automatic or userManaged.
+	// Ref: https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets
+	// +optional
+	// +immutable
+	Replication *Replication `json:"replication,omitempty"`
+}
+
+// Replication policy that defines the replication configuration of data.
+type Replication struct {
+	// The replication policy for this secret.
+	// This devaites from the internal representation of the google API since it's an interface.
+	// Ref: google/cloud/secretmanager/v1/resources.proto
+	// There are two possible values : usermanaged and automatic
+	// +optional
+	// +immutable
+	ReplicationType *ReplicationType `json:"ReplicationType"`
+}
+
+// ReplicationType refers to type of replication of the secrets
+type ReplicationType struct {
+	// +optional
+	// +immutable
+	// +nullable
+	Automatic *ReplicationAutomatic `json:"ReplicationAutomatic"`
+
+	// +optional
+	// +immutable
+	UserManaged *ReplicationUserManaged `json:"ReplicationUserManaged"`
+}
+
+// ReplicationAutomatic has fields for automatic replication of secrets
+type ReplicationAutomatic struct {
+	// +optional
+	// +immutable
+	// +nullable
+}
+
+// ReplicationUserManaged has fields for user managed replication of secrets
+type ReplicationUserManaged struct {
+
+	// Locations of the secret.
+	// +immutable
+	Replicas []*ReplicationUserManagedReplica `json:"Replicas"`
+}
+
+// ReplicationUserManagedReplica has fields of each replica of user managed replica type
+type ReplicationUserManagedReplica struct {
+	// Location of the secret.
+	// +immutable
+	Location string `json:"location,omitempty"`
```

Comment:
Should be pointer type if optional, though if it is the only field in this embedded struct we could consider making it required and just the parent struct optional.
```suggestion
	Location *string `json:"location,omitempty"`
```

Reply:
Yes, making the struct optional reflects what GCP expects. If user has started filling the struct, it means `Location` is a required field. Also, newer GCP SDK does have additional fields